### DISCUSSION TITLE
SOLVED: Extending `burnSteps` case setting to be a vector

### DISCUSSION THREAD

**Discussion by keckler** [terrapower/armi#561] [Ideas]
From my understanding, right now the `burnSteps` case setting is just a single value, which is applied to all cycles. This is fine for most cases, but sometimes one might be simulating a reactor with cycle lengths that vary quite a bit, and different cycles would call for different numbers of burn steps. Or you want to examine the isotopics in detail after a burnup, so you'd want to have a fine time mesh for the decay cycles but not for the actual burn cycles. I could propose other use cases as well.

Would anybody have any thoughts on extending the `burnSteps` setting to allow for multiple values? This would be in line with the `cycleLength`/`cycleLengths` or `powerFractions` case settings.

**Comment by jakehader** [comment]
I think that this would be a great addition/improvement for the framework. I wonder if it makes sense to also define the number of days for the burnup steps too rather than just the number of steps themselves. For example, cycle length is 365 days, 3 burnup steps, with the burnup step times being 10 days, 55 days, and 300 days rather than just being a uniform 121.6667 days each (365/3).

I think that this support in the framework could help with sensitivity studies on burnup, help with modeling specific benchmarks, and help to add further depletion support for LWRs.

**Reply by keckler** [reply]
I definitely agree. It didn't even register to me that such a capability doesn't exist, but you're right that it is very important to have for thermal systems, in addition to many other potential use cases.

**Reply by jakehader** [reply]
This was actually an internal ask within our team too, so I think that we should definitely do this. 

So as long as there aren't any issues from @ntouran or @john-science on this I suggest:

- Convert burnSteps into a list.
- Add a list of lists for burnStep times, where the index on the top list is the cycle number and an inner list should have number of days for the burn steps that matches with the int defined in burnSteps for that cycle (i.e, if 3 burnSteps are given then there needs to be a list of length 3 for the days). 

Setting days might be more convenient to think in for the burnStep times, but we then should also check that the number of days set by the user matches the cycle length. 

Now that I am writing this out, I think we need to be careful with availability factors too and understand how those impact the burnSteps. Are the burnSteps defined from start of cycle to end of cycle or are they defined from start of cycle to (end of cycle - nonavailability time). If the latter then this should be checked.

Also, this brings up another thought on power fractions. It seems that we ought to be able to set power fractions on the burnSteps rather than only on the cycle level. There are definitely use cases in real reactors or in analysis where at different times in the cycle the power is changed. Right now I believe the power fractions are not as granular as they should be.

**Reply by keckler** [reply]
I agree that this could get a little more complicated than it might first appear, and it probably calls for a unified plan between a couple different changes to be implemented.

Just spit-balling here, but something like this is what I would personally like to see.
First of all, I think this type of input could be complicated enough to warrant its own 'blueprints' file. If the user only wanted to do simple steps like what we have now, they could use the existing case settings. But if they wanted to be more granular, they could specify the name of a file on a new case setting `cycle data: my_cycle_data.yaml`. Inside this cycle data file would look like:
```
cycles:
  # cycle 0
   - cumulative days: [1, 5, 10, 100]
     powerFractions: [0.1, 0.5, 1, 1]
  # cycle 1
  - cumulative days: [1, 5, 6, 15, 100]
    powerFractions: [0.1, 0.5, 0.7, 1.0, 0.9]
```

I also think it would be very helpful to introduce an optional cycle name attribute that gets logged in ARMI so that the cycles can have some context outside of just their index. Something like:
```
cycles:
  - cycle name: "Cycle 1 (initial power tests) "
    cumulative days: [1, 5, 10, 100]
    powerFractions: [0.1, 0.5, 1, 1]
  - cycle name: "Cycle 2.a" 
    cumulative days: [1, 5, 6, 15, 50]
    powerFractions: [0.1, 0.5, 0.7, 1.0, 0.9]
  - cycle name: "Cycle 2.b" 
    cumulative days: [50]
    powerFractions: [1.0]
```

I also think would could offer a `step days` option in addition to the `cumulative days`. 

We would have to specify something about how the powerFractions are interpolated (I assume we would force them to be constant over the step). 

In addition, I think we could offer some possibility in there for specifying a finer mesh on the post-cycle decay period that is defined by `availabilityFactors`. I think right now that is just one step, with no way to extract information at specific decay times (without breaking the decay steps out into a separate cycle).

**Reply by jakehader** [reply]
Lots of things here, but I personally like it. I think that being able to allow users to assign custom state point names rather than just having cycle 0, time node 0 etc. could be a good idea. Also, I sort of like the idea of changing the implementation of these settings to be under a "cycles" data structure. This would be a breaking change to a lot of inputs and we'd want to be cleaver with making a migration script to support this change too. We have examples of this within the code base so it's not an issue. Although, there will be some changes needed to our internal GUI to streamline handling this type of input structure.

I like the idea of having multiple decay times after shutdown for decay heat calculations, or even other post shutdown analyses with time (i.e., source range detection analysis).

**Reply by keckler** [reply]
Thanks for the feedback. 

I actually don't know if this would need to be a breaking change. I imagine that we could make the various case settings mutually-exclusive with each other such that the user could either provide their cycle input in the old format _or_ supply the new data structure. So existing inputs would continue to work.

**Reply by john-science** [reply]
Sorry, I just realized I hadn't commented on this.

I like the idea. It has a clear purpose and benefit.

I can't think of any particularly tricky problems implementing this. If there is some issue that would make this a big task and not a small one, I can't think of it at the moment.

**Reply by keckler** [reply]
I can put some effort into implementing this in the next week or two.

**Reply by keckler** [reply]
Okay, well I take it back. This will have to be a breaking change, from what I can tell. 

It is not going to break the user's cases, but rather I think it will break a bunch of plugins, primarily because `burnSteps` is stored as a case setting and is always accessed through the `self.cs["burnSteps"]` syntax. In order to make it variable, it would need to be stored on the operator and accessed via `self.r.p.burnSteps` syntax instead, so that the current cycle can be known. This is how `powerFractions` and `availabilityFactors` are stored and accessed right now.

So I don't know exactly the best way to go about doing this right now...

**Reply by ntouran** [reply]
I think vectorizing burnSteps to be consistent with cycleLengths, powerFractions, and availabilityFactors makes all the sense in the world. Those each have two settings (e.g. `cycleLength` and `cycleLengths`) that allow users to input one unchanging one or a cycle-wise vector. In both cases, a list with values for each cycle is created and stored on the operator. 

For this case, I guess keeping the existing setting would be convenient for not breaking backwards compatibility of input files, and we'd want to make an additional setting for providing a vector input.  Ideas:

* `burnStepsByCycle`?

And yes you'll have to `:s/self.cs["burnSteps"]/self.r.p.burnSteps` everywhere that uses it. 

Along these lines, while making this new parameter and setting, would it be worth attempting to make it a better name? burnSteps is a bit unconventional. Perhaps subcycles? You could actually do this while maintaining compatibility with old inputs:

New settings:
* `numSubcycles`
* `numSubcyclesByCycle`

new reactor parameters:
* `subcycleNum`

New setting migration:
* map value from `burnSteps` to `numSubcycles`

Regarding days vs. numbers, that is probably convenient. I recommend that be considered as a separate discussion and development. That could be handled as just an alternate setting that can be used mutually exclusively with e.g. numSubcycles, etc.

**Comment by keckler** [comment]
This has been resolved: #637
